compiler_opt/rl/regalloc/gin_configs/network.gin

```diff
@@ -9,6 +9,8 @@ regalloc.config.get_observation_processing_layer_creator.with_z_score_normalizat
 create_agent.policy_network = @regalloc_network.RegAllocNetwork
 
 RegAllocNetwork.preprocessing_combiner=@tf.keras.layers.Concatenate()
-RegAllocNetwork.fc_layer_params=(80, 40)
+num_units_layer1=80
```

Comment:
this change hurts the flexibility, suppose we want to run behavioral cloning of (40, 20) and (40, 20, 20), this is not going to work.

Is it possible to iterate between tuples? so [(40, 20), (40, 20, 20)]

Reply:
It's not possible to directly iterate between tuples as the bayesian optimizer only wants to operate over single continuous (well, discrete for the integers, but I'm guessing it's floating inside the implementation) variables. It's definitely not particularly flexible, but since we can't add extra variables to the optimizer at runtime, the only other solutions that I can think of are pretty clunky and also somewhat inflexible (eg something like having a `num_layers` attribute and being able to set the maximum number of layers and just having extra variables for each per-layer parameter that are sometimes ignored depending upon the value of `num_layers`). Kind of clunky and the optimizer there would be working on variables that don't always have an impact. Shouldn't be a massive deal, but it will make the acquisition function less efficient.

Reply:
does the bayesian optimizer supports some kind of category choice? we can treat it as categories instead of numerical

Reply:
There is a category choice available (documentation [here](https://keras.io/api/keras_tuner/hyperparameters/#choice-method)), but tuples are not on the officially supported list of types that it will cycle through. I could implement as an integer/choice of an integer and then just map that integer to the choices made available in the gin config, but that's still a pretty clunky solution.

Reply:
ic, in your config you can have [(80, 40), (40, 20, 20)] for it. For numerical stuff, you can have [2, 3].

In your code, you can check whether the content is an int or float or bool, if they are, keep them, or else convert them to string. The code does not look too clunky in this way i guess?

Reply:
Yeah. Probably shouldn't be too bad. Converting everything to a string would probably make a decently clean implementation. I'll work on an implementation and see what I can come up with.

Reply:
Just wrote up a preliminary implementation of this. I didn't end up using the choice option as if we want to have more than 1 per layer hyperparameter, it quickly becomes less feasible. The keras tuner package allows for dynamically grabbing hyperparameters per trial, so depending upon the trial it just asks for different parameters from the optimizer. Might be slightly less efficient for the optimizer, but it's probably better than writing our own stuff to concatenate all reasonable permutations of per layer hparams and piping that in as a choice for the model.